lib/kdl/include/kdl/result.h

```diff
@@ -0,0 +1,669 @@
+/*
+ Copyright 2020 Kristian Duske
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+ documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+ rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ persons to whom the Software is furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+ Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+ WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef result_h
+#define result_h
+
+#include "kdl/overload.h"
+#include "kdl/result_forward.h"
+
+#include <absl/types/optional.h>
+#include <absl/types/variant.h>
+
+namespace kdl {
+    /**
+     * Wrapper class that can contain either a value or one of several errors.
+     *
+     * An instance of this class represents an expectation for the result of applying a function if that function
+     * returns either a value or throws an error.
+     *
+     * A result is considered successful if it contains a value, and a failure if it contains an error.
+     *
+     * @tparam Value the type of the value
+     * @tparam Errors the types of the possible errors
+     */
+    template <typename Value, typename... Errors>
+    class [[nodiscard]] result {
+    public:
+        using value_type = Value;
+    private:
+        absl::variant<value_type, Errors...> m_value;
+    private:
+        /**
+         * Creates a result that wraps the given value, which can be a value or an error.
+         *
+         * If the value or error is passed by rvalue reference, it is moved into this result.
+         *
+         * @tparam T the type of the given value or error
+         * @param t the value or error to wrap
+         */
+        template <typename T>
+        result(T&& t)
+        : m_value(std::forward<T>(t)) {}
+    public:
+        template <typename V, typename std::enable_if<std::is_convertible_v<V, Value>>::type* = nullptr>
+        static result success(V&& v) {
+            return result(std::forward<V>(v));
```

Comment:
One more thing, how do you know this is going to construct the success case? (e.g. assume all the variant's types have default constructors.) Does it try them in order?

`result(value_type(std::forward<V>(v)));`

would eliminate any doubt I think. We need to also say in the docs that you can't use reuse your Value type as an error type.

Reply:
Using the same type more than once in the argument list of `result` yields a compile error when you try to construct it. I'll try to add more SFINAE type checking to give bettter error messages tho.

Reply:
`success()` should pass `std::in_place_index_t<0>()` as the first argument to the variant constructor to force it to actually create the success case.

Making `error()` fully safe is trickier, I was just playing around on godbolt and got this:

```
#include <variant>
#include <string>
#include <type_traits>
#include <cassert>

template<class Value, class... Errors>
struct result {

    using VariantType = std::variant<Value, Errors...>;

    template <typename V>
    static VariantType make_value(V&& v) {
        // force creation of index 0 (Value)
        return VariantType(std::in_place_index_t<0>(), v);
    }

    template <typename V>
    static VariantType make_error(V&& v) {
        // check that V is one of Errors
        static_assert(std::disjunction_v<std::is_same<V, Errors>...>);

        return VariantType(std::in_place_type_t<V>(), v);
    }
};

int main() {

    // implicit conversion from float to int   
    auto v1 = result<int, float, std::string>::make_value(1.2f);
    assert(v1.index() == 0);

    auto v2 = result<int, float, std::string>::make_error(1.3f);
    assert(v2.index() == 1);

    auto v3 = result<int, float, std::string>::make_error(std::string("abc"));
    assert(v3.index() == 2);

    // compile error: make_error() wants float or std::string
    //auto v4 = result<int, float, std::string>::make_error(1);

    return 0;
}
```

Reply:
Ok, I have added similar code to our factory methods.

Reply:
Nice, I didn't know about `std::disjunction` yet -- I was looking for something like that the other day ;-)